### PR TITLE
W-14203139-Concurrent_Buffer_Implementation

### PR DESCRIPTION
**Pull request by juanfloresMule**
_Hi, thank you for your work. We understand that you want to merge your changes and move on from this issue, but we also want to maintain the safety, readability, and testability of our code. Because of this, we kindly ask that you confirm that you have fulfilled the prerequisites for each category of activity before merging your changes._

### If the Pull Request has a dependency update:

- [ ] I have read the Release Notes for the new version (and intermediate versions if the jump would include more than one). _Don't blindly trust that the dependency will honor SEMVER, always read carefully the release notes_
- [ ] Java 8+ support is maintained
- [ ] The new version has no vulnerabilities
- [ ] A team member reviewed the Pull Request.
- [ ] No backward compatibility is broken.
- [ ] If it is a "provided" dependency, it has been checked that if it is suggested, the version number is the same.

### For ALL the Releases:

- [ ] There is a task in GUS for this change.
- [ ] The corresponding Release Notes have been written and shared with the documentation team.
- [ ] The new code has tests and they have before and after methods to be sure that you are working with a clean environment and that you are leaving the environment clean after they finish.
- [ ] If you are using reflection, create a test to call the methods you are invoking, this way, if there is a change in the API, the test will be able to detect it.
- [ ] If the project has a parent pom and modules, the release pipeline only uploads the modules, so, please, make sure that the parent was deployed in the repositories before considering this release finished.
- [ ] Notify in the Slack Channel that the release is complete, so the Docs team can merge the Documentation and Release Notes PR. The docs team always checks in Anypoint Exchange that the connector is released before merging the release notes.
- [ ] Add the connector build to GUS, and assign that build to the GUS ticket.
- [ ] Create a change case request, based on the Change Case Management doc https://docs.google.com/document/d/1tMJlTTZLaXMLiYwxlMBFY1yJwBmejhc4-IP8HAXgDfY/edit#

**NOTE**: _(applies only for Core-Connectors connectors and modules) The release process ends when you merge the pull request that updates the support branch to the new snapshot, please don't forget to do this._

### Patch Version:

- [ ] The Pull Request has been peer-reviewed.
- [ ] No backward compatibility is broken.
- [ ] Documentation: (Required) Release Notes PR with compatibility table. Share it with the Docs team.

### Minor Version:

- [ ] Have a document with the specification of the release.
- [ ] Create a slack channel (eg: rl-conninfra-sftp-1.1.1-new-chiper) to coordinate the release of this version with the documentation team and the architects of our team (at least one week before the release). Share the spec doc with the docs team.
- [ ] The documentation fork has been done and has been updated with the changes related to the new functionalities.
- [ ] The pull request has been reviewed by a peer and the team leader.
- [ ] No backward compatibility is broken.
- [ ] Public Documentation: Minor releases nearly always require a revision of the documentation because the release usually includes new features as well as bug fixes. Even if there are no user-facing changes, documentation has to be versioned for every minor release.
  - [ ] (Required) Release Notes PR with compatibility table. Share it with the Docs team.
  - [ ] (Likely required) Reference guide. Run the script and share the file with the Docs team.
  - [ ] (Assess) Often a minor release impacts UI and examples, so the documentation (user guides, examples, troubleshooting guides) should be reviewed and updated accordingly. Notify the docs team if any updates are needed.

### Major Version:

- [ ] Have a document with the specification of the release.
- [ ] Create a slack channel (eg: rl-conninfra-sftp-2.0.0) to coordinate the release of this version with the documentation team and the architects of our team (at least one month before the release). Share the spec doc with the docs team.
- [ ] The documentation fork has been made and updated with the changes related to the new functionalities and possible compatibility problems with the previous version.
- [ ] The systems properties have been removed, detailing in the documentation the changes in the default behavior of the product.
- [ ] The TODO comments of the code have been reviewed.
- [ ] A colleague, the team leader, and the team architect have reviewed the Pull Request.
- [ ] Public Documentation: Major releases break backward compatibility and should never be released without documentation, as this breaks the user experience.
  - [ ] (Required) Upgrade and Migration Guide: Every major release must have a detailed upgrade guide (template) with information about what was changed and what steps the user has to take to ensure the connector works as expected. Share the details with the Docs team so they can update the guide.
  - [ ] (Required) Release Notes PR with compatibility table. Share it with the docs team.
  - [ ] (Required) Reference guide. Run the script and share the file with the Docs team.
  - [ ] User guide updates, new screenshots or examples (if applicable), share the updates with the docs team.

### CODE REVIEW

File: src/main/java/org/mule/extension/sftp/internal/connection/SftpConcurrentClient.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+public class SftpConcurrentClient extends DefaultSftpClient {
+
+  /**
+   * @param clientSession          The {@link ClientSession}
+   * @param initialVersionSelector The initial {@link SftpVersionSelector} - if {@code null} then version 6 is
+   *                               assumed.
+   * @param errorDataHandler       The {@link SftpErrorDataHandler} to handle incoming data through the error stream
+   *                               - if {@code null} the data is silently ignored
+   * @throws IOException If failed to initialize
+   */
```
Comment (acurci-at-mulesoft) [inline src/main/java/org/mule/extension/sftp/internal/connection/SftpConcurrentClient.java:28]: Move this javadoc over the right method

File: src/main/java/org/mule/extension/sftp/internal/connection/SftpConcurrentClient.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.apache.sshd.client.session.ClientSession;
+import org.apache.sshd.common.util.buffer.Buffer;
+import org.apache.sshd.sftp.client.SftpErrorDataHandler;
+import org.apache.sshd.sftp.client.SftpVersionSelector;
+import org.apache.sshd.sftp.client.impl.DefaultSftpClient;
+
+public class SftpConcurrentClient extends DefaultSftpClient {
```
Comment (acurci-at-mulesoft) [inline src/main/java/org/mule/extension/sftp/internal/connection/SftpConcurrentClient.java:19]: Add a Javadoc to explain the reason for creating this class and don't forget to mention the GUS number of the issue that starts this problem

File: src/main/java/org/mule/extension/sftp/internal/connection/SftpConcurrentClient.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  private final Lock sendLock = new ReentrantLock();
+
+  public SftpConcurrentClient(ClientSession clientSession, SftpVersionSelector initialVersionSelector,
+                              SftpErrorDataHandler errorDataHandler)
+      throws IOException {
+    super(clientSession, initialVersionSelector, errorDataHandler);
+  }
+
+  @Override
+  public int send(int cmd, Buffer buffer) throws IOException {
```
Comment (acurci-at-mulesoft) [inline src/main/java/org/mule/extension/sftp/internal/connection/SftpConcurrentClient.java:38]: Remember to add a TODO to think about the possibility of improving this method to use a pool (or a map) of buffers

File: src/main/java/org/mule/extension/sftp/internal/connection/SftpClient.java [hunk 4]
```diff
@@ -204,7 +204,7 @@ private void checkExists(String path) {
   }
 
   private void connect() throws IOException {
-    org.apache.sshd.sftp.client.SftpClientFactory scf = org.apache.sshd.sftp.client.SftpClientFactory.instance();
+    SftpConcurrentClientFactory scf = SftpConcurrentClientFactory.INSTANCE;
```
Comment (acurci-at-mulesoft) [inline src/main/java/org/mule/extension/sftp/internal/connection/SftpClient.java:207]: You could add the instance() method to preserve the way that it gets to that instance.

**Review by acurci-at-mulesoft — CHANGES_REQUESTED** [review]

File: src/main/java/org/mule/extension/sftp/internal/connection/SftpConcurrentClient.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+   * @throws IOException If failed to initialize
+   */
+  public SftpConcurrentClient(ClientSession clientSession, SftpVersionSelector initialVersionSelector,
+                              SftpErrorDataHandler errorDataHandler)
+      throws IOException {
+    super(clientSession, initialVersionSelector, errorDataHandler);
+  }
+
+  /**
+   * TODO: The solution of synchronizing the send operation can cause a bottleneck
```
Comment (acurci-at-mulesoft) [inline src/main/java/org/mule/extension/sftp/internal/connection/SftpConcurrentClient.java:44]: Please, could you create the user story for this issue and add the gus number in the TODO

**Review by acurci-at-mulesoft — DISMISSED** [review]

File: src/main/java/org/mule/extension/sftp/internal/connection/SftpClient.java (hunk 4 above)
Comment (juanfloresMule) [reply src/main/java/org/mule/extension/sftp/internal/connection/SftpClient.java]: Done

File: src/main/java/org/mule/extension/sftp/internal/connection/SftpConcurrentClient.java (hunk 3 above)
Comment (juanfloresMule) [reply src/main/java/org/mule/extension/sftp/internal/connection/SftpConcurrentClient.java]: Done

File: src/main/java/org/mule/extension/sftp/internal/connection/SftpConcurrentClient.java (hunk 2 above)
Comment (juanfloresMule) [reply src/main/java/org/mule/extension/sftp/internal/connection/SftpConcurrentClient.java]: Done

File: src/main/java/org/mule/extension/sftp/internal/connection/SftpConcurrentClient.java (hunk 5 above)
Comment (juanfloresMule) [reply src/main/java/org/mule/extension/sftp/internal/connection/SftpConcurrentClient.java]: Done

**Review by sabecasismulesoft — DISMISSED** [review]

**Review by acurci-at-mulesoft — CHANGES_REQUESTED** [review]

File: src/test/munit/file-listener-test-cases/listen-multiple-files.xml [hunk 6]
```diff
@@ -0,0 +1,59 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<mule xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:munit="http://www.mulesoft.org/schema/mule/munit"
+      xmlns:munit-tools="http://www.mulesoft.org/schema/mule/munit-tools"
+      xmlns="http://www.mulesoft.org/schema/mule/core"
+      xmlns:sftp="http://www.mulesoft.org/schema/mule/sftp"
+      xmlns:java="http://www.mulesoft.org/schema/mule/java"
```
[the remark this answers is not in the document]
Comment (acurci-at-mulesoft) [reply src/test/munit/file-listener-test-cases/listen-multiple-files.xml:7]: (I am talking about java namespace.)

File: src/test/munit/file-listener-test-cases/listen-multiple-files.xml [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+                <fixed-frequency frequency="3000"/>
+            </scheduling-strategy>
+        </sftp:listener>
+        <set-variable value="#[%dw 2.0&#10;output application/java&#10;---&#10;attributes.path]" doc:name="Set Variable" variableName="filePath"/>
+        <set-variable value="#[%dw 2.0&#10;output application/java&#10;---&#10;attributes.size]" doc:name="Set Variable" variableName="fileSize"/>
+        <set-variable value="#[%dw 2.0&#10;output application/java&#10;---&#10;attributes.fileName]" doc:name="Set Variable" variableName="fileName"/>
+        <logger level="ERROR" doc:name="ERROR: Initial vars" message="Initial vars: #[vars]"/>
+        <logger level="ERROR" doc:name="ERROR: Start File processing" message="[file: #[vars.fileName]][size: #[vars.fileSize]] Start File Processing" />
+        <logger level="ERROR" doc:name="ERROR: File content"  message="[file: #[vars.fileName]][size: #[vars.fileSize]] WMS file content: #[payload]"/>
+        <logger level="ERROR" doc:name="ERROR: Completed File Processing" message="[file: #[vars.fileName]][size: #[vars.fileSize]] File processing was completed. The file will be deleted from the server..."/>
```
[the remark this answers is not in the document]
Comment (juanfloresMule) [reply src/test/munit/file-listener-test-cases/listen-multiple-files.xml]: changed to level="INFO"

File: src/test/munit/file-listener-test-cases/listen-multiple-files.xml (hunk 6 above)
[the remark this answers is not in the document]
Comment (juanfloresMule) [reply src/test/munit/file-listener-test-cases/listen-multiple-files.xml]: Done

File: src/test/munit/file-listener-test-cases/listen-multiple-files.xml [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+      xmlns="http://www.mulesoft.org/schema/mule/core"
+      xmlns:sftp="http://www.mulesoft.org/schema/mule/sftp"
+      xmlns:java="http://www.mulesoft.org/schema/mule/java"
+      xmlns:doc="http://www.mulesoft.org/schema/mule/documentation"
+      xsi:schemaLocation="
+		http://www.mulesoft.org/schema/mule/core http://www.mulesoft.org/schema/mule/core/current/mule.xsd
+		http://www.mulesoft.org/schema/mule/munit http://www.mulesoft.org/schema/mule/munit/current/mule-munit.xsd
+		http://www.mulesoft.org/schema/mule/munit-tools  http://www.mulesoft.org/schema/mule/munit-tools/current/mule-munit-tools.xsd
+        http://www.mulesoft.org/schema/mule/sftp http://www.mulesoft.org/schema/mule/sftp/current/mule-sftp.xsd
+        http://www.mulesoft.org/schema/mule/java http://www.mulesoft.org/schema/mule/java/current/mule-java.xsd">
```
[the remark this answers is not in the document]
Comment (juanfloresMule) [reply src/test/munit/file-listener-test-cases/listen-multiple-files.xml]: done

**Review by acurci-at-mulesoft — APPROVED** [review]

**Pull request merged**